libbeat/common/transport/client.go

```diff
@@ -0,0 +1,253 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package transport
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/testing"
+)
+
+type Client struct {
+	dialer  Dialer
+	network string
+	host    string
+	config  *Config
+
+	conn  net.Conn
+	mutex sync.Mutex
+}
+
+type Config struct {
+	Proxy   *ProxyConfig
+	TLS     *TLSConfig
+	Timeout time.Duration
+	Stats   IOStatser
+}
+
+func MakeDialer(c *Config) (Dialer, error) {
```

Comment:
I'd be in favor of passing configuration objects by value. By value also makes it clear that we own a copy, no matter if you modify the original config.

Reply:
Makes sense, will fix!

Reply:
I'm going to address all feedback related to the transport refactoring in a separate PR that will resolve https://github.com/elastic/beats/issues/16728.

Reply:
Please see https://github.com/elastic/beats/pull/16734.